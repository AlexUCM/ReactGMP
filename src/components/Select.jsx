import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { useClickOutside } from '../hooks/useClickOutside';
import styled from 'styled-components';

const StyledSelect = styled.div`
  position: relative;
  height: 24px;
  width: 150px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0 0 0 24px;
  color: #fff;
  cursor: pointer;

  &:after {
    content: '\\25BC';
    margin: 0 0 0 10px;
    font-size: 12px;
    color: #f65261;
  }

  ${(props) =>
    props.isOpen &&
    `
    &:after {
      transform: rotate(180deg);
  }
  `}
`;

const Dropdown = styled.div`
  position: absolute;
  display: none;
  right: 0;
  top: 26px;
  width: 174px;
  background-color: #232323;
  border: 1px solid #424242;
  border-radius: 4px;
  z-index: 2;

  ${(props) =>
    props.isOpen &&
    `
    display: block;
  `}
`;

const SortValue = styled.p`
  line-height: 32px;
  padding: 0 8px;

  ${(props) =>
    props.isActive &&
    `
    background-color: #f65261;
  `}
`;

export const Select = ({ value, toggleSortValue }) => {
  const sortingValues = { RELEASE_DATE: 'RELEASE DATE', RATING: 'RATING' };

  const ref = useRef(null);

  const [isOpen, setIsOpen] = useState(false);

  useClickOutside(ref, () => isOpen && setIsOpen(false));

  return (
    <StyledSelect
      onClick={() => setIsOpen((state) => !state)}
      isOpen={isOpen}
      ref={ref}
    >
      {value}
      <Dropdown onClick={toggleSortValue} isOpen={isOpen}>
        {Object.values(sortingValues).map((val) => {
          return (
            <SortValue key={val} isActive={val === value}>
              {val}
            </SortValue>
          );
        })}
      </Dropdown>
    </StyledSelect>
  );
};

Select.propTypes = {
  value: PropTypes.string.isRequired,
  toggleSortValue: PropTypes.func.isRequired,
};
