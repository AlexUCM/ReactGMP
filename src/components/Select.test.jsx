import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/react';
import { Select } from './Select';

describe('Select', () => {
  const getComponent = ({
    value = 'RELEASE DATE',
    toggleSortValue = jest.fn(),
  }) => render(<Select value={value} toggleSortValue={toggleSortValue} />);
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot();
  });
  it('Dropdown should be visible', () => {
    const { getByTestId } = getComponent({});
    expect(getByTestId('dropdown')).toHaveStyle('display: none');
    fireEvent.click(getByTestId('select'));
    expect(getByTestId('dropdown')).toHaveStyle('display: block');
    fireEvent.click(document);
    expect(getByTestId('dropdown')).toHaveStyle('display: none');
  });
});
