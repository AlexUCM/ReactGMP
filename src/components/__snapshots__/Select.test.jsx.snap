// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ joLNfS"
    data-testid="select"
  >
    RELEASE DATE
    <div
      class="sc-gsTCUz hsgYzw"
      data-testid="dropdown"
    >
      <p
        class="sc-dlfnbm hDlHYy"
        data-testid="RELEASE DATE"
      >
        RELEASE DATE
      </p>
      <p
        class="sc-dlfnbm hCPypq"
        data-testid="RATING"
      >
        RATING
      </p>
    </div>
  </div>
</DocumentFragment>
`;
